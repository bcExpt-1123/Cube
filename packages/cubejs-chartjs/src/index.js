import cubejs from 'cubejs-client';

class ChartjsResultSet {
  constructor(resultSet) {
    this.resultSet = resultSet;
  }

  timeSeries(config) {
    return {
      type: 'line',
      data: {
        datasets: this.resultSet.series()
          .map(s => ({ label: s.name, data: s.series.map(r => ({ t: r[0], y: r[1] }) ) }) )
      },
      options: {
        scales: {
          xAxes: [{
            type: 'time',
            unit: this.resultSet.query().timeDimensions[0].granularity,
            distribution: 'series',
            bounds: 'data'
          }]
        }
      },
      ...config
    }
  }
}

cubejs.chartjs = (resultSet) => {
  return new ChartjsResultSet(resultSet);
};

export default cubejs;