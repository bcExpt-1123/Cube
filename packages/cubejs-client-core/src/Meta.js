import {unnest, fromPairs} from 'ramda';

const memberMap = (memberArray) => fromPairs(memberArray.map(m => [m.name, m]));

export default class Meta {
  constructor(metaResponse) {
    this.meta = metaResponse;
    const { cubes } = this.meta;
    this.cubes = cubes;
    this.cubesMap = fromPairs(cubes.map(c => [
      c.name,
      { measures: memberMap(c.measures), dimensions: memberMap(c.dimensions), segments: memberMap(c.segments) }
    ]));
  }

  membersForQuery(query, memberType) {
    return unnest(this.cubes.map(c => c[memberType]));
  }

  resolveMember(memberName, memberType) {
    const [cube] = memberName.split('.');
    if (!this.cubesMap[cube]) {
      return { title: memberName, error: `Cube not found ${cube} for path '${memberName}'` };
    }
    const member = this.cubesMap[cube][memberType][memberName];
    if (!member) {
      return { title: memberName, error: `Path not found '${memberName}'` };
    }
    return member;
  }
}
