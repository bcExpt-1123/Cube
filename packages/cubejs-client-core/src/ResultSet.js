import { groupBy, pipe, toPairs, uniq, flatten, map } from 'ramda';

export default class ResultSet {
  constructor(loadResponse) {
    this.loadResponse = loadResponse;
  }

  series(pivotConfig) {
    return this.seriesNames(pivotConfig).map(({ title, key }) => ({
      title,
      series: this.pivotedRows(pivotConfig).map(({ category, ...obj }) => ({ value: obj[key], category }))
    }));
  }

  axisValues(axis) {
    const query = this.loadResponse.query;
    return row => {
      const value = (measure) =>
        axis.filter(d => d !== 'measures')
        .map(d => row[d]).concat(measure ? [measure] : [])
        .map(v => v != null ? v : 'âˆ…').join(', ');
      if (axis.find(d => d === 'measures') && (query.measures || []).length) {
        return query.measures.map(value);
      }
      return [value()];
    };
  }

  normalizePivotConfig(pivotConfig) {
    const query = this.loadResponse.query;
    let timeDimensions = (query.timeDimensions || []).filter(td => !!td.granularity);
    pivotConfig = pivotConfig || timeDimensions.length ? {
      x: timeDimensions.map(td => td.dimension),
      y: query.dimensions || []
    } : {
      x: query.dimensions || [],
      y: []
    };
    if (!pivotConfig.x.concat(pivotConfig.y).find(d => d === 'measures')) {
      pivotConfig.y = pivotConfig.y.concat(['measures']);
    }
    return pivotConfig;
  }

  static measureFromAxis(axis) {
    const axisValues = axis.split(', ');
    return axisValues[axisValues.length - 1];
  };

  pivotedRows(pivotConfig) {
    // TODO missing date filling
    pivotConfig = this.normalizePivotConfig(pivotConfig);
    return pipe(groupBy(this.axisValues(pivotConfig.x)), toPairs)(this.loadResponse.data).map(([category, rows]) => ({
      category,
      ...(rows.map(row => this.axisValues(pivotConfig.y)(row).map(series => {
            let measure = pivotConfig.x.find(d => d === 'measures') ?
              ResultSet.measureFromAxis(category) :
              ResultSet.measureFromAxis(series);
            return {
              [series]: row[measure] && Number.parseFloat(row[measure])
            }
          }).reduce((a, b) => Object.assign(a, b), {})
        )).reduce((a, b) => Object.assign(a, b), {})
      })
    );
  }

  totalRow() {
    return this.pivotedRows()[0];
  }

  categories(pivotConfig) { //TODO
    return this.pivotedRows(pivotConfig);
  }

  seriesNames(pivotConfig) {
    pivotConfig = this.normalizePivotConfig(pivotConfig);
    return pipe(map(this.axisValues(pivotConfig.y)), uniq, flatten)(this.loadResponse.data).map(axis => ({
      title: pivotConfig.y.find(d => d === 'measures') ? axis.replace(
        ResultSet.measureFromAxis(axis),
        this.loadResponse.annotation.measures[ResultSet.measureFromAxis(axis)].title
      ) : axis,
      key: axis
    }))
  }

  query() {
    return this.loadResponse.query;
  }

  rawData() {
    return this.loadResponse.data;
  }
}