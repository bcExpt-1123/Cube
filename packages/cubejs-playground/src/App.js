import React, { Component } from 'react';
import "antd/dist/antd.css";
import "./index.css";
import { Layout} from "antd";
import cubejs from '@cubejs-client/core';
import { fetch } from 'whatwg-fetch';
import PlaygroundQueryBuilder from './PlaygroundQueryBuilder';

const { Header, Content } = Layout;

const API_URL = ``;

class App extends Component {
  constructor(props) {
    super(props);
    this.state = {};
  }

  cubejsApi() {
    if (!this.cubejsApiInstance && this.state.cubejsToken) {
      this.cubejsApiInstance = cubejs(this.state.cubejsToken, {
        apiUrl: this.state.apiUrl + '/cubejs-api/v1'
      });
    }
    return this.cubejsApiInstance;
  }

  async componentDidMount() {
    const res = await fetch(API_URL + '/playground/context');
    const result = await res.json();
    this.setState({
      cubejsToken: result.cubejsToken,
      apiUrl: result.apiUrl
    });
    window.analytics && window.analytics.identify(result.anonymousId);
  }

  render() {
    return [
      <Layout>
        <Header>
          <img src='./cubejs-logo.svg' style={{ display: 'inline', width: 50}} />
          <h2 style={{ color: '#fff', display: 'inline' }}>Cube.js Playground</h2>
        </Header>
        <Content style={{ padding: '25px', margin: '25px' }}>
          {this.cubejsApi() && <PlaygroundQueryBuilder
            query={{
              measures: ['Orders.count'],
              dimensions: ['Orders.status']
            }}
            cubejsApi={this.cubejsApi()}
            apiUrl={this.state.apiUrl}
            cubejsToken={this.state.cubejsToken}
          />}
        </Content>
      </Layout>
    ];
  }
}

export default App;
