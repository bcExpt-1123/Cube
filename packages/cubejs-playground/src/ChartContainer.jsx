import React from 'react';
import { Card, Button } from 'antd';
import { getParameters } from 'codesandbox-import-utils/lib/api/define';
import { map } from 'ramda';
import PrismCode from './PrismCode';
import { playgroundAction } from './events';

class ChartContainer extends React.Component {
  constructor(props) {
    super(props);
    this.state = { showCode: false };
  }

  render() {
    const { resultSet, error, codeExample, render, title, codeSandboxSource, dependencies } = this.props;

    const { getParameters } = require('codesandbox-import-utils/lib/api/define');

    const parameters = getParameters({
      files: {
        'index.js': {
          content: codeSandboxSource,
        },
        'package.json': {
          content: {
            dependencies: {
              'react-dom': 'latest',
              ...map(() => 'latest', dependencies)
            }
          },
        },
      },
      template: 'create-react-app'
    });

    const codeSandboxLink = `https://codesandbox.io/api/v1/sandboxes/define?parameters=${parameters}`;

    const extra =
      (<Button.Group>
        <Button
          onClick={() => {
            playgroundAction('Show Code');
            this.setState({ showCode: !this.state.showCode });
          }}
          icon="code"
          size="small"
          type={this.state.showCode ? 'primary' : 'default'}
        >
          {this.state.showCode ? 'Hide Code' : 'Show Code'}
        </Button>
        <Button
          href={codeSandboxLink}
          target="_blank"
          icon="code-sandbox"
          size="small"
          onClick={() => playgroundAction('Open Code Sandbox')}
        >
          Edit
        </Button>
      </Button.Group>);

    return (<Card
      title={title}
      style={{ minHeight: 420 }}
      extra={extra}
    >
      {this.state.showCode ? <PrismCode code={codeExample} /> : render({ resultSet, error })}
    </Card>);
  }
}

export default ChartContainer;
