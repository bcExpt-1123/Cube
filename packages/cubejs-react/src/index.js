import QueryRenderer from './QueryRenderer.jsx';

export { QueryRenderer };
