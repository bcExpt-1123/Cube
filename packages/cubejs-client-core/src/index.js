import { fetch } from 'whatwg-fetch';
import ResultSet from './ResultSet';
import SqlQuery from './SqlQuery';
import ProgressResult from './ProgressResult';

const API_URL = process.env.CUBEJS_API_URL;

class CubejsApi {
  constructor(apiToken, options) {
    options = options || {};
    this.apiToken = apiToken;
    this.apiUrl = options.apiUrl || API_URL;
  }

  request(url, config) {
    return fetch(
      `${this.apiUrl}${url}`,
      Object.assign({ headers: { Authorization: this.apiToken, 'Content-Type': 'application/json' }}, config || {})
    )
  }

  loadMethod(request, toResult, options, callback) {
    if (typeof options === 'function' && !callback) {
      callback = options;
      options = undefined;
    }

    options = options || {};


    const loadImpl = async () => {
      const response = await request();
      if (response.status === 502) {
        return loadImpl(); // TODO backoff wait
      }
      const body = await response.json();
      if (body.error === 'Continue wait') {
        if (options.progressCallback) {
          options.progressCallback(new ProgressResult(body));
        }
        return loadImpl();
      }
      if (response.status !== 200) {
        throw new Error(body.error); // TODO error class
      }
      return toResult(body);
    };
    if (callback) {
      loadImpl().then(r => callback(null, r), e => callback(e));
    } else {
      return loadImpl();
    }
  }

  load(query, options, callback) {
    return this.loadMethod(
      () => this.request(`/load?query=${JSON.stringify(query)}`),
      (body) => new ResultSet(body),
      options,
      callback
    );
  }

  sql(query, options, callback) {
    return this.loadMethod(
      () => this.request(`/sql?query=${JSON.stringify(query)}`),
      (body) => new SqlQuery(body),
      options,
      callback
    );
  }
}

export default (apiToken) => {
  return new CubejsApi(apiToken);
};