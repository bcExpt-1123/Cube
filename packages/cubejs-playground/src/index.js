import React from 'react';
import ReactDOM from 'react-dom';
import { Router, Route } from 'react-router-dom';
import createHistory from 'history/createHashHistory';
import IndexPage from './IndexPage';
import ExplorePage from './ExplorePage';
import SchemaPage from './SchemaPage';
import App from './App';
import { page } from './events';

const history = createHistory({ basename: process.env.PUBLIC_URL });
history.listen((location) => {
  page(location);
});

ReactDOM.render(
  <Router history={history}>
    <App>
      <Route
        exact
        path="/"
        component={IndexPage}
      />
      <Route
        path="/explore"
        component={ExplorePage}
      />
      <Route
        path="/schema"
        component={SchemaPage}
      />
    </App>
  </Router>,
  // eslint-disable-next-line no-undef
  document.getElementById('root')
);
