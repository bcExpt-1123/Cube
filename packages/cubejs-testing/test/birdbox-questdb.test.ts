import { createBirdBoxTestCase } from './events-only-test-case';
import { startBirdBoxFromContainer } from '../src';

createBirdBoxTestCase('questdb', () => startBirdBoxFromContainer({
  name: 'questdb',
  loadScript: 'questdb-load-events.sh',
}));
