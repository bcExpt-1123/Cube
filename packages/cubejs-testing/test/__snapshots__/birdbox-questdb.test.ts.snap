// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`questdb HTTP Transport #1 Events.count: #1 Events.count 1`] = `
Array [
  Object {
    "Events.count": "999",
  },
]
`;

exports[`questdb HTTP Transport #2 Events.count with Events.type order by Events.count DESC: #2 Events.count with Events.type order by Events.count DESC 1`] = `
Array [
  Object {
    "Events.count": "513",
    "Events.type": "PushEvent",
  },
  Object {
    "Events.count": "120",
    "Events.type": "CreateEvent",
  },
  Object {
    "Events.count": "104",
    "Events.type": "IssueCommentEvent",
  },
  Object {
    "Events.count": "92",
    "Events.type": "WatchEvent",
  },
  Object {
    "Events.count": "57",
    "Events.type": "IssuesEvent",
  },
  Object {
    "Events.count": "32",
    "Events.type": "PullRequestEvent",
  },
  Object {
    "Events.count": "21",
    "Events.type": "ForkEvent",
  },
  Object {
    "Events.count": "21",
    "Events.type": "GollumEvent",
  },
  Object {
    "Events.count": "21",
    "Events.type": "PullRequestReviewCommentEvent",
  },
  Object {
    "Events.count": "14",
    "Events.type": "DeleteEvent",
  },
  Object {
    "Events.count": "2",
    "Events.type": "ReleaseEvent",
  },
  Object {
    "Events.count": "1",
    "Events.type": "MemberEvent",
  },
  Object {
    "Events.count": "1",
    "Events.type": "CommitCommentEvent",
  },
]
`;

exports[`questdb WS Transport #1 Events.count: #1 Events.count 1`] = `
Array [
  Object {
    "Events.count": "999",
  },
]
`;

exports[`questdb WS Transport #2 Events.count with Events.type order by Events.count DESC: #2 Events.count with Events.type order by Events.count DESC 1`] = `
Array [
  Object {
    "Events.count": "513",
    "Events.type": "PushEvent",
  },
  Object {
    "Events.count": "120",
    "Events.type": "CreateEvent",
  },
  Object {
    "Events.count": "104",
    "Events.type": "IssueCommentEvent",
  },
  Object {
    "Events.count": "92",
    "Events.type": "WatchEvent",
  },
  Object {
    "Events.count": "57",
    "Events.type": "IssuesEvent",
  },
  Object {
    "Events.count": "32",
    "Events.type": "PullRequestEvent",
  },
  Object {
    "Events.count": "21",
    "Events.type": "ForkEvent",
  },
  Object {
    "Events.count": "21",
    "Events.type": "GollumEvent",
  },
  Object {
    "Events.count": "21",
    "Events.type": "PullRequestReviewCommentEvent",
  },
  Object {
    "Events.count": "14",
    "Events.type": "DeleteEvent",
  },
  Object {
    "Events.count": "2",
    "Events.type": "ReleaseEvent",
  },
  Object {
    "Events.count": "1",
    "Events.type": "MemberEvent",
  },
  Object {
    "Events.count": "1",
    "Events.type": "CommitCommentEvent",
  },
]
`;

exports[`questdb responseFormat http+responseFormat=compact option#1+2: result-type 1`] = `
Array [
  Object {
    "Events.count": "513",
    "Events.type": "PushEvent",
  },
  Object {
    "Events.count": "120",
    "Events.type": "CreateEvent",
  },
]
`;

exports[`questdb responseFormat http+responseFormat=compact option#1: result-type 1`] = `
Array [
  Object {
    "Events.count": "513",
    "Events.type": "PushEvent",
  },
  Object {
    "Events.count": "120",
    "Events.type": "CreateEvent",
  },
]
`;

exports[`questdb responseFormat http+responseFormat=compact option#2: result-type 1`] = `
Array [
  Object {
    "Events.count": "513",
    "Events.type": "PushEvent",
  },
  Object {
    "Events.count": "120",
    "Events.type": "CreateEvent",
  },
]
`;

exports[`questdb responseFormat http+responseFormat=default: result-type 1`] = `
Array [
  Object {
    "Events.count": "513",
    "Events.type": "PushEvent",
  },
  Object {
    "Events.count": "120",
    "Events.type": "CreateEvent",
  },
]
`;

exports[`questdb responseFormat ws+responseFormat=compact option#1+2: result-type 1`] = `
Array [
  Object {
    "Events.count": "513",
    "Events.type": "PushEvent",
  },
  Object {
    "Events.count": "120",
    "Events.type": "CreateEvent",
  },
]
`;

exports[`questdb responseFormat ws+responseFormat=compact option#1: result-type 1`] = `
Array [
  Object {
    "Events.count": "513",
    "Events.type": "PushEvent",
  },
  Object {
    "Events.count": "120",
    "Events.type": "CreateEvent",
  },
]
`;

exports[`questdb responseFormat ws+responseFormat=compact option#2: result-type 1`] = `
Array [
  Object {
    "Events.count": "513",
    "Events.type": "PushEvent",
  },
  Object {
    "Events.count": "120",
    "Events.type": "CreateEvent",
  },
]
`;

exports[`questdb responseFormat ws+responseFormat=default: result-type 1`] = `
Array [
  Object {
    "Events.count": "513",
    "Events.type": "PushEvent",
  },
  Object {
    "Events.count": "120",
    "Events.type": "CreateEvent",
  },
]
`;
